inc/Engine/Common/JobManager/APIHandler/AbstractSafeAPIClient.php

```diff
@@ -0,0 +1,149 @@
+<?php
+
+namespace WP_Rocket\Engine\Common\JobManager\APIHandler;
+
+use WP_Error;
+
+/**
+ * Class AbstractSafeAPIClient
+ *
+ * This abstract class provides a base for making API requests.
+ * It includes methods for sending GET and POST requests, and handles transient caching to mitigate the impact of API failures.
+ *
+ * @package WP_Rocket\Engine\Common\JobManager\APIHandler
+ */
+abstract class AbstractSafeAPIClient {
+
+	/**
+	 * Get the transient key.
+	 *
+	 * @return string The transient key.
+	 */
+	abstract protected function get_transient_key();
+
+	/**
+	 * Get the API URL.
+	 *
+	 * @return string The API URL.
+	 */
+	abstract protected function get_api_url();
+
+	/**
+	 * Send a GET request.
+	 *
+	 * @param array $params The request parameters.
+	 * @param bool  $safe Send safe request WP functions or not, default to not.
+	 *
+	 * @return mixed The response from the API.
+	 */
+	public function send_get_request( $params = [], $safe = false ) {
+		return $this->send_request( 'GET', $params, $safe );
+	}
+
+	/**
+	 * Send a POST request.
+	 *
+	 * @param array $params The request parameters.
+	 * @param bool  $safe Send safe request WP functions or not, default to not.
+	 *
+	 * @return mixed The response from the API.
+	 */
+	public function send_post_request( $params = [], $safe = false ) {
+		return $this->send_request( 'POST', $params, $safe );
+	}
+
+	/**
+	 * Send a request to the API.
+	 *
+	 * @param string $method The HTTP method (GET or POST).
+	 * @param array  $params The request parameters.
+	 * @param bool   $safe Send safe request WP functions or not, default to not.
+	 * @return mixed The response from the API, or WP_Error if a timeout is active.
+	 */
+	private function send_request( $method, $params = [], $safe = false ) {
+		$api_url = $this->get_api_url();
+
+		if ( get_transient( $this->get_transient_key() . '_timeout_active' ) ) {
+			return new WP_Error( 429, __( 'Too many requests.', 'rocket' ) );
+		}
+
+		$params['method'] = strtoupper( $method );
+
+		$response = $this->send_remote_request( $api_url, $method, $params, $safe );
+
+		if ( is_wp_error( $response ) ) {
+			$this->set_timeout_transients();
+			return $response;
+		}
+
+		$body = wp_remote_retrieve_body( $response );
+		if ( empty( $body ) || ( is_array( $response ) && ! empty( $response['response']['code'] ) && 200 !== $response['response']['code'] ) ) {
+			$this->set_timeout_transients();
+			return new WP_Error( 500, __( 'Not valid response.', 'rocket' ) );
+		}
```

Comment:
In case the return is not a 200, I am afraid this will be occulting completely the response for potential debug: there will be no way to know what was actually returned as we erase the response code and just send back a WP_Error 500?
Could we either:
- add a log, or
- Use the response code in the WP_Error code, or
- Past the response code in the message of the WP_Error?

Reply:
To do later on